tasks/tasks.go

```diff
@@ -0,0 +1,577 @@
+	if err == nil {
+		os.Remove(testFile)
+	}
+	if err != nil || !bytes.Equal(readTestData, writtenTestData) {
+		return fmt.Errorf("Could not read from given data directory!")
+	}
+	return nil
+}
+
+// global variables for managing tasks
+var running bool                // true if tasks are processing, false if not
+var taskChannels channelsType   // channels used for processing tasks
+var stopHeartbeat chan struct{} // send a pulse to this channel to halt polling
+
+// Starts processing tasks according to the given configuration, returning an
+// informative error if anything prevents this.
+func Start() error {
+	if running {
+		return fmt.Errorf("Tasks are already running and cannot be started again.")
+	}
+
+	// does the directory exist and is it writable/readable?
+	err := validateDataDirectory(config.Service.DataDirectory)
+	if err != nil {
+		return err
+	}
+
+	// can we access the local endpoint?
+	_, err = endpoints.NewEndpoint(config.Service.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	// allocate channels
+	taskChannels = channelsType{
+		TaskId: make(chan uuid.UUID, 32),
+		Task:   make(chan taskType, 32),
+		Status: make(chan TransferStatus, 32),
+		Error:  make(chan error, 32),
+		Poll:   make(chan struct{}),
+		Stop:   make(chan struct{}),
+	}
+
+	// start processing tasks
+	go processTasks()
+
+	// start the polling heartbeat
+	slog.Info(fmt.Sprintf("Task statuses are updated every %d ms",
+		config.Service.PollInterval))
+	pollInterval := time.Duration(config.Service.PollInterval) * time.Millisecond
+	go heartbeat(pollInterval, taskChannels.Poll)
+
+	// okay, we're running now
+	running = true
+
+	return nil
+}
+
+// Stops processing tasks. Adding new tasks and requesting task statuses are
+// disallowed in a stopped state.
+func Stop() error {
+	taskChannels.Stop <- struct{}{}
+	err := <-taskChannels.Error
+	running = false
+	return err
+}
+
+// Returns true if tasks are currently being processed, false if not.
+func Running() bool {
+	return running
+}
+
+// Adds a new transfer task associated with the user with the specified Orcid ID
+// to the manager's set, returning a UUID for the task. The task is defined by
+// specifying the names of the source and destination databases and a set of
+// file IDs associated with the source.
+func Add(orcid, source, destination string, fileIDs []string) (uuid.UUID, error) {
+	var taskId uuid.UUID
+
+	// verify that we can fetch the task's source and destination databases
+	// without incident
+	_, err := databases.NewDatabase(orcid, source)
+	if err != nil {
+		return taskId, err
+	}
+	_, err = databases.NewDatabase(orcid, destination)
+	if err != nil {
+		return taskId, err
+	}
+
+	// create a new task and send it along for processing
+	taskChannels.Task <- taskType{
+		Orcid:       orcid,
+		Source:      source,
+		Destination: destination,
+		FileIds:     fileIDs,
+	}
+	select {
+	case taskId = <-taskChannels.TaskId:
+	case err = <-taskChannels.Error:
+	}
```

Comment:
weird formatting!

Reply:
Agreed, and it's the same for `switch` (which at least doesn't have C's silly fallthrough behavior to worry about!)